### ISSUE TITLE
Bug: [no-base-to-string] false positive when `String` is shadowed in an outer scope

### ISSUE THREAD

**Issue by zaewc** [typescript-eslint/typescript-eslint#12490]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.
- [x] If I used AI assistance to prepare this issue, I have reviewed and followed the [AI Contribution Policy](https://typescript-eslint.io/contributing/ai-policy/).

### Playground Link

https://typescript-eslint.io/play/#ts=6.0.3&fileType=.tsx&code=MYewdgzgLgBAylATgSzAcxgXhgCgB4BcMArmANZggDuYAlFgHwwDkEAhgGYCmzA3AFD8OpYFGTgYHHPQDe-GPCSo0OGQF9aAtfyA&eslintrc=N4KABBYEQE4K4BsCmBnKAuMpKSgAQBcBPAB1QGMYBLEggWlQSoDsCB6ZgezoCMBDFEjoFuKAtWYBzDNCQwYnGFHCQAviFVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
const String = (x: unknown) => 'safe';
function f() {
  String({});
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  parserOptions: { project: true },
  rules: {
    "@typescript-eslint/no-base-to-string": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

No error

`String` is shadowed by the user-defined function above, so the call isn't the global String constructor and shouldn't be checked

For comparison, declaring the same `const String` in the same scope as the call is correctly ignored, it only breaks when the call sits in a nested scope below the shadowing declaration

### Actual Result

'{}' will use Object's default stringification format ('[object Object]') when stringified. 4:10 - 4:12

**Comment by kirkwaiblinger** [comment]
Interesting, seems something is broken with the scope logic in 

https://github.com/typescript-eslint/typescript-eslint/blob/2df540cd8bbeb3e2c56d516912f69bf63c1e9450/packages/eslint-plugin/src/rules/no-base-to-string.ts#L301-L314
